enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+---
+title: platform-operators
+authors:
+  - "@timflannagan"
+reviewers:
+  - TBD
+approvers:
+  - "@joelanford"
+  - "@bparees"
+  - TBD
+api-approvers:
+  - TBD
+creation-date: 2022-06-30
+last-updated: 2022-07-06
+tracking-link:
+  - https://issues.redhat.com/browse/OLM-2232
+  - https://issues.redhat.com/browse/OLM-2513
+  - https://issues.redhat.com/browse/OLM-2605
+see-also:
+  - https://issues.redhat.com/browse/OLM-2447
+  - https://issues.redhat.com/browse/OLM-2570
+  - https://issues.redhat.com/browse/CORS-1873
+  - https://issues.redhat.com/browse/OCPPLAN-7589
+  - https://github.com/openshift/enhancements/blob/master/enhancements/update/cvo-techpreview-manifests.md
+replaces: []
+superseded-by: []
+---
+
+# Platform Operators
+
+## Summary
+
+The enhancement outlines a mechanism that enable support for the lifecycle of platform operators. A platform operator (PO) is an OLM-based cluster extension that can be installed during day 0 operations, and lifecycles with the cluster. This mechanism will ultimately be divided into phases, and this proposal covers the first phase is more detail.
+
+This first phase introduces a basic PO installation mechanism, and helps build the foundational layer for subsequent phases. Future phases may introduce a first-class PO management mechanism for the OLM component.
+
+## Motivation
+
+The motivation for platform operators is two-fold. The strategic direction for OpenShift is to make the core installation smaller by default and at the same time make it more flexible so that customers can tailor their installations to their use cases. This is what we’re calling “Composable OpenShift”.
+
+To make the core smaller, OpenShift is attempting to reduce the CVO-managed payload to the very minimum set of operators required to run the OpenShift control plane. From a control plane perspective, any non-essential operators can be installed after the cluster has been installed (or upgraded). However, from a customer standpoint, their business needs may dictate that these non-essential operators are still included in the lifecycle of cluster installations and upgrades. The Platform Operator (PO) concept will allow cluster administrators to choose which of these non-control-plane-essential operators to include in the cluster lifecycle.
+
+Secondly, some portion of the layered product operators in the OpenShift catalogs are conducive with inclusion in the cluster lifecycle (e.g. cert-manager, service binding operator, and the local storage operator). The Platform Operator concept will open the door for more of OLM’s optional operators to be included in (and impact) the cluster lifecycle. This will allow customers to further tailor their cluster installs and upgrades to their business needs.
+
+### User Stories
+
+- As a cluster admin, I want to be able to install OLM-based Operators at cluster creation, and participate in the cluster's lifecycle.
```

Comment:
Yes, spelling those things out would be helpful.

Reply:
I'll add this to the summary section in the next round of edits.

Reply:
> automatically upgrading to new versions of the PO when the cluster is being upgraded

Thinking about this some more: do we need to define this more granularity such that we outlined what happens in the following scenarios:

- What happens when the cluster's minor version is upgraded?
- What happens when the cluster's patch version is upgraded?
- Do we need to care about cluster rollbacks? Are rollbacks supported behaviors in OCP?

I think what I've heard before is that when the cluster's minor version is upgraded, we'll automatically upgrade the POs packages that are currently installed. _But_ we've also said that we'd like to enable PO teams to release on their own cadence, without being tied to the OCP payload lifecycle, and provide admins with control over approving out-of-band upgrades.

Maybe this is something we can punt on until phase 1, but I think we'd want to document and express those scenarios here to set the stage longer term.

> blocking cluster upgrades when the PO can't tolerate the new version (or has another reason to block the upgrade)

This is slightly (if you squint) related: in the case of a cluster minor version upgrade, when do POs get upgraded? Does this process of upgrading POs happen after the control plane components are upgraded? Is it sufficient to upgrade all POs packages all at once, or is there some sort of CVO run-level style hierarchy we'd need to expose for a layered rollout of packages?

Reply:
> * What happens when the cluster's minor version is upgraded?

I would expect the platform operators to be updated to be compatible with that minor version.

> * What happens when the cluster's patch version is upgraded?

I don't think I would expect any platform operators to be upgraded in this case. Expressing compatibility with patch versions of OCP is too finicky.

> * Do we need to care about cluster rollbacks? Are rollbacks supported behaviors in OCP?

Good questions.

> I think what I've heard before is that when the cluster's minor version is upgraded, we'll automatically upgrade the POs packages that are currently installed. _But_ we've also said that we'd like to enable PO teams to release on their own cadence, without being tied to the OCP payload lifecycle, and provide admins with control over approving out-of-band upgrades.

*As a cluster admin, I want to be able to upgrade a platform operator I have installed independently of OpenShift so I can receive bug fixes in the platform operator.*

*As a cluster admin, I want my platform operators to stay compatible with the version of OpenShift I am running I upgrade my cluster so the cluster continues to work smoothly after the upgrade.*

"Stay compatible" doesn't automatically imply an upgrade, but in practice it probably will require that most of the time.

> Maybe this is something we can punt on until phase 1, but I think we'd want to document and express those scenarios here to set the stage longer term.
> 
> > blocking cluster upgrades when the PO can't tolerate the new version (or has another reason to block the upgrade)
> 
> This is slightly (if you squint) related: in the case of a cluster minor version upgrade, when do POs get upgraded? Does this process of upgrading POs happen after the control plane components are upgraded? Is it sufficient to upgrade all POs packages all at once, or is there some sort of CVO run-level style hierarchy we'd need to expose for a layered rollout of packages?

I'd say keep it simple to start out and do them all at once.

It's generally going to be easier for a given version of an operator to support versions N-1 and N of OpenShift, because supporting N and N+1 would mean predicting the future (API removals or changes, etc.). So, I would say to go from OpenShift N-1 to N, upgrade the platform operators first, then upgrade OpenShift.

Reply:
> I don't think I would expect any platform operators to be upgraded in this case. 

I disagree on this one.  Any time the cluster is being upgraded, we should be upgrading the platform operators to the latest version that is compatible w/ the cluster version (typically defined by the minor version).  It doesn't matter if it's a patch or minor version upgrade.

The point here being, if i'm at 4.13.3 and my PO is at 1.2.3, and now i upgrade my cluster to 4.13.4, and there's a PO 1.2.4 available, i want you to upgrade that too while i'm taking my upgrade maintenance window.

That's what "lifecycle the operators w/ the cluster means" - make the POs behave like COs in terms of when they upgrade.

> Expressing compatibility with patch versions of OCP is too finicky.

There are really only two possibilities here:

1) we allow POs to specify their OCP dep version all the way down to the patch level (they don't have to, but we allow it).  In this case, we won't upgrade the PO anyway, unless the new version is compatible w/ the new OCP version
2) we don't allow POs to specify dep beyond the minor version.  Regardless this is what i'd expect most POs to do.  If we do this, then nothing constrains what PO version you get when you install fresh on a new OCP cluster of a given patch level anyway, so there's no reason not to also allow upgrading of the PO version.

(to be clear i think we should be on option (1), but discourage operators from specifying patch level dependencies unless they have a good reason (such as kernel/rhcos deps like CNV has)

> It's generally going to be easier for a given version of an operator to support versions N-1 and N of OpenShift, because supporting N and N+1 would mean predicting the future (API removals or changes, etc.). So, I would say to go from OpenShift N-1 to N, upgrade the platform operators first, then upgrade OpenShift.

ultimately this also gets into our desire to have preflight checking.  But absent that, the answer in my mind is "both"

You may need to upgrade the PO before you upgrade the cluster (in the case that the PO's maxOCPVersion is less than the incoming OCP version).  (Note that this gets slightly tricky because effectively it means the POM should report upgradeable=true, based on the fact that while the cluster can't/shouldn't be upgraded as is, we know there is a PO upgrade that will(may?) resolve that block.  This will require some careful orchestration w/ the CVO)

You may also be able to(if not required to) upgrade the PO after you upgrade the cluster (this goes back to what i said above about upgrading everything while you're taking a maintenance window), so if newer versions of the PO are now available to the cluster immediately after the upgrade(really "during" the upgrade), we should be consuming them.

> Do we need to care about cluster rollbacks? Are rollbacks supported behaviors in OCP?

we support rollbacks to a previous patch level, but not to a previous minor.  And it's only supported as a workaround until you can upgrade again, it's not intended that you downgrade and then stay there long term.  For a first pass we may be able to exempt POs from this, but ultimately if downgrading the cluster means we need an older version of the PO (for dependency reasons), we should be able to rerun the dep resolution and "upgrade" to the older PO version (i.e. rukpak and deppy should allow "upgrading" from a newer PO version to an older one, the same way OCP allows you to "upgrade" to an older payload)

Reply:
> > I don't think I would expect any platform operators to be upgraded in this case.
> 
> I disagree on this one. Any time the cluster is being upgraded, we should be upgrading the platform operators to the latest version that is compatible w/ the cluster version (typically defined by the minor version). It doesn't matter if it's a patch or minor version upgrade.

I could go along with that more readily if we limited it to patch versions.

> 
> The point here being, if i'm at 4.13.3 and my PO is at 1.2.3, and now i upgrade my cluster to 4.13.4, and there's a PO 1.2.4 available, i want you to upgrade that too while i'm taking my upgrade maintenance window.
> 
> That's what "lifecycle the operators w/ the cluster means" - make the POs behave like COs in terms of when they upgrade.

Yes, that's reasonable. I was concerned about automatically upgrading to 1.3.0 automatically in that case, and I'm not sure we want that. It would need to be compatible to support upgrading the PO as part of a move from 4.13 to 4.14 but still. Expressing my desire to upgrade the cluster to a new patch level implies that I don't want new features.

> > It's generally going to be easier for a given version of an operator to support versions N-1 and N of OpenShift, because supporting N and N+1 would mean predicting the future (API removals or changes, etc.). So, I would say to go from OpenShift N-1 to N, upgrade the platform operators first, then upgrade OpenShift.
> 
> ultimately this also gets into our desire to have preflight checking. But absent that, the answer in my mind is "both"
> 
> You may need to upgrade the PO before you upgrade the cluster (in the case that the PO's maxOCPVersion is less than the incoming OCP version). (Note that this gets slightly tricky because effectively it means the POM should report upgradeable=true, based on the fact that while the cluster can't/shouldn't be upgraded as is, we know there is a PO upgrade that will(may?) resolve that block. This will require some careful orchestration w/ the CVO)
> 
> You may also be able to(if not required to) upgrade the PO after you upgrade the cluster (this goes back to what i said above about upgrading everything while you're taking a maintenance window), so if newer versions of the PO are now available to the cluster immediately after the upgrade(really "during" the upgrade), we should be consuming them.

Both are possible to implement. I think we'll make life easier for everyone (OLM developers, operator developers, release managers, cluster admins) if we pick an order and always do it the same way. Unless you think there's ever a case where a PO couldn't be backwards compatible?

Reply:
> Both are possible to implement. I think we'll make life easier for everyone (OLM developers, operator developers, release managers, cluster admins) if we pick an order and always do it the same way. Unless you think there's ever a case where a PO couldn't be backwards compatible?

upgrading POs after/during the cluster upgrade is definitely "simpler" than upgrading them beforehand, because upgrading them beforehand requires us to determine, at the time we know the admin is starting to think about upgrading, whether or not there is a newer version of the PO that is compatible w/ the intended upgrade version.  And in reality we don't want the admin to find out during their maintenance weekend that there isn't one.

So requiring POs be compatible w/ the next minor avoids that problem.  But of course that's also impossible to guarantee.

Suppose i'm at ocp 4.13 and i install a PO v1.1.1 that's compatible with v4.13 and v4.14.

Now I upgrade to v4.14, but as v4.15 doesn't exist yet, there is no newer version of the PO that exists yet and support v4.15.

Now when i go to upgrade to v4.15 I will need to first upgrade my PO to a newer PO that supports v4.15 (assuming one even exists).  And i need to do that *before* i upgrade my cluster, otherwise i'll be running a PO on v4.15 when that PO doesn't claim to support v4.15.

Ultimately I think the pragmatic thing to do here is probably to make the following assertions:

1) If your currently installed PO explicitly is not compatible with the next anticipated minor (or even the next anticipated z), upgradeable=false and the admin must take action to manually upgrade the PO first.
2) post any (patch or minor) OCP upgrade(as part of the OCP upgrade really), we will upgrade all your POs to the newest available/compatible versions based on the new OCP level.

This leaves a few situations where in theory we could do more for you, instead in the hands of admins, but at least for a first pass I think it strikes the right balance.

To your point about this meaning we might upgrade a PO to a new minor version as part of an OCP patch level upgrade...yes that is true, but if the PO declares compatibility I don't think that's entirely unreasonable.  

If we want to get really fancy I suppose we could go down a path of:

1) during ocp patch level upgrades, only allow POs to upgrade to a new patch level
2) during ocp minor level upgrades, allow POs to upgrade to a new minor level

I do kind of like that constraint.

Reply:
> (i.e. rukpak and deppy should allow "upgrading" from a newer PO version to an older one, the same way OCP allows you to "upgrade" to an older payload)

IIRC, this should be on OLM's 1.x roadmap but there's still some ongoing discussion around how to handle this scenario best given it's easy to get workloads into bad states. At the base rukpak layer, our current stance is that component is very WYSIWYG, which means that instead of using the term "upgrading" we defer to using "pivoting". In this context, "pivoting" is essentially telling rukpak controllers to progress to the desired bundle content without any embedding any significant preflight checks, and instead relies on higher level components (e.g. deppy) to correctly configure the rukpak APIs.

When deppy is performing dependency resolution, and a new resolution outcome is produced, then rukpak will simply honor that outcome with the assumption that aggregate set of bundle contents can be successfully persisted to the cluster. With all of this said, we should be inheriting this behavior from OLM's 1.x vision, but we may require downstream safeguards to help reduce the change for getting workloads in bad states.

---

This is all good discussion. I updated the EP, and clarified the "cluster lifecycle" phrasing. For phase 0, I marked upgrades as out-of-scope, and therefore blocking cluster upgrades for POs would also be out-of-scope. That would mean for that this first phase, we're really only interested in ensuring that a failed, individual PO rollout would act like a SLO today, and the overall cluster rollout would fail.

Reply:
> 1. during ocp patch level upgrades, only allow POs to upgrade to a new patch level
> 2. during ocp minor level upgrades, allow POs to upgrade to a new minor level

What about allowing POs to upgrade to a new major level? We'll need a story for that, and I have a hard time seeing the answer being "wait until OCPv5".

But this seems right to me conceptually and it would definitely make the "support PO downgrades during cluster patch version downgrades" story more palatable.

This also highlights (if it wasn't already obvious) the fact that the PO controller will need to add/remove constraints into resolution automatically based on the steady vs. upgrading state of the cluster. It'll essentially be:
1. Steady state: pin to range `=X.Y.Z`
2. Patch upgrade: pin to range `>=X.Y.Z <X.Y+1.0-0`
3. Minor upgrade : pin to range `>=X.Y.Z <X+1.0.0-0`

But during steady state, we need to permit manual upgrades triggered by cluster admins. In that case perhaps, we need something like an optional `spec.steadyStateVersionRange` field on the PO that allows the cluster admin to override the default "pin to this exact version" behavior?